src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Object.WriteTests.cs

```diff
@@ -146,5 +148,79 @@ public static void WriteObjectWithNumberHandling()
             var options = new JsonSerializerOptions { NumberHandling = JsonNumberHandling.AllowReadingFromString };
             JsonSerializer.Serialize(new object(), options);
         }
+
+        /// <summary>
+        /// This test is constrained to run on Windows and MacOSX because it causes
+        /// problems on Linux due to the way deferred memory allocation works. On Linux, the allocation can
+        /// succeed even if there is not enough memory but then the test may get killed by the OOM killer at the
+        /// time the memory is accessed which triggers the full memory allocation.
+        /// Also see <see cref="Utf8JsonWriterTests.WriteLargeJsonToStreamWithoutFlushing"/>
+        /// </summary>
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
+        [ConditionalFact(nameof(Utf8JsonWriterTests.IsX64))]
+        [OuterLoop]
+        public static void SerializeLargeListOfObjects()
+        {
+            Dto dto = new()
+            {
+                Prop1 = int.MaxValue,
+                Prop2 = int.MinValue,
+                Prop3 = "AC",
+                Prop4 = 500,
+                Prop5 = int.MaxValue / 2,
+                Prop6 = 250.5M / 3,
+                Prop7 = 250.5M / 3,
+                Prop8 = 250.5M / 3,
+                Prop9 = 250.5M / 3,
+                Prop10 = 250.5M / 3,
+                Prop11 = 150.6M / 3,
+                Prop12 = 150.6M / 3,
+                Prop13 = DateTimeOffset.Now,
+                Prop14 = DateTimeOffset.Now,
+                Prop15 = DateTimeOffset.Now,
+                Prop16 = DateTimeOffset.Now,
+                Prop17 = 3,
+                Prop18 = DateTime.Now,
+                Prop19 = DateTime.Now,
+                Prop20 = 25000,
+                Prop21 = DateTime.Now
+            };
+
+            List<Dto> items = Enumerable.Repeat(dto, 3_000_000).ToList();
+
+            // Try-catch is used since OOM might not occur on all machines.
```

Comment:
which machines does it not occur? you said this is not running on Linux and on Windows/OSX you'd presumably get OOM?

Reply:
@steveharter mentioned he was able to serialize 2M elements (https://github.com/dotnet/runtime/issues/73992#issuecomment-1217112231) whereas I OOM'd in one of my iterations. I assume a similar `win-x64` was used so I wrote it this way to be safe.

Reply:
Yeah this should be `Assert.Throws`.